### PR TITLE
Fix Pupil course contact and others, rework to typed GQL

### PR DESCRIPTION
**Pull request by Jonasdoubleyou**

### CODE REVIEW

File: src/pages/pupil/SingleCoursePupil.tsx [hunk 1]
```diff
@@ -33,16 +34,16 @@ function OtherParticipants({ subcourseId }: { subcourseId: number }) {
         { variables: { subcourseId } }
     );
 
-    if (loading) return <CenterLoadingSpinner />;
+    if (loading || !data) return <CenterLoadingSpinner />;
```
Comment (Jonasdoubleyou) [inline src/pages/pupil/SingleCoursePupil.tsx:37]: loading implies !data, better to make this explicit to make data non nullable

File: src/pages/pupil/SingleCoursePupil.tsx [hunk 2]
```diff
@@ -222,7 +221,7 @@ const SingleCoursePupil = () => {
         },
     ];
 
-    if (subcourse?.isInstructor || subcourse?.isParticipant) {
+    if (subcourse?.isParticipant) {
```
Comment (Jonasdoubleyou) [inline src/pages/pupil/SingleCoursePupil.tsx:224]: pupil can never be instructor

File: src/pages/pupil/SingleCoursePupil.tsx [hunk 3]
```diff
@@ -236,23 +235,23 @@ const SingleCoursePupil = () => {
     return (
         <WithNavigation headerTitle={course?.name.substring(0, 20)} showBack isLoading={loading} headerLeft={<NotificationAlert />}>
             <Stack space={space['2']} paddingX={space['1.5']}>
-                <SubcourseData course={course} subcourse={subcourse} isInPast={isInPast} />
+                {course && subcourse && <SubcourseData course={course} subcourse={subcourse} isInPast={isInPast} />}
```
Comment (Jonasdoubleyou) [inline src/pages/pupil/SingleCoursePupil.tsx:238]: course and subcourse could be null here, though the component is not rendered then due to the isLoading={loading} above. But again better to make explicit I guess

File: src/pages/pupil/SingleCoursePupil.tsx [hunk 4]
```diff
@@ -236,23 +235,23 @@ const SingleCoursePupil = () => {
...
                         subcourse={subcourse}
                         canJoinSubcourse={canJoinData?.subcourse?.canJoin}
-                        joinedSubcourse={joinedSubcourseData}
-                        joinedWaitinglist={joinedWaitinglist}
-                        leftSubcourseData={leftSubcourseData}
-                        leftWaitinglist={leftWaitinglist}
+                        joinedSubcourse={joinedSubcourseData?.subcourseJoin}
+                        joinedWaitinglist={joinedWaitinglist?.subcourseJoinWaitinglist}
+                        leftSubcourseData={leftSubcourseData?.subcourseLeave}
+                        leftWaitinglist={leftWaitinglist?.subcourseLeaveWaitinglist}
```
Comment (Jonasdoubleyou) [inline src/pages/pupil/SingleCoursePupil.tsx:254]: I'm pretty sure that this did not work before? @LomyW @timidak

File: src/pages/pupil/single-course/PupilCourseButtons.tsx [hunk 5]
```diff
@@ -131,12 +131,12 @@ const PupilCourseButtons: React.FC<ActionButtonProps> = ({
                         <WaitinglistBanner courseStatus={courseTrafficStatus} onLeaveWaitinglist={setLeaveWaitingslistModal} loading={loadingWaitinglistLeft} />
                     </VStack>
                 )}
-                {subcourse.isParticipant && (
+                {subcourse.isParticipant && subcourse.canContactInstructor.allowed && (
```
Comment (Jonasdoubleyou) [inline src/pages/pupil/single-course/PupilCourseButtons.tsx:134]: The actual bug: The course contact button is shown although participants might not be able to contact the instructor (instructors can turn this off for courses)

File: src/pages/pupil/single-course/PupilCourseButtons.tsx [hunk 6]
```diff
@@ -131,12 +131,12 @@ const PupilCourseButtons: React.FC<ActionButtonProps> = ({
...
                     <Button variant="outline" onPress={() => setShowMessageModal(true)}>
                         {t('single.actions.contactInstructor')}
                     </Button>
                 )}
-                {subcourse?.isParticipant && subcourse?.published && <JoinMeeting subcourse={subcourse} refresh={refresh} />}
+                {subcourse?.isParticipant && <JoinMeeting subcourse={subcourse} refresh={refresh} />}
```
Comment (Jonasdoubleyou) [inline src/pages/pupil/single-course/PupilCourseButtons.tsx:139]: pupils can only see published courses anyways

**Review by LomyW — APPROVED** [review]
like it! thanks for the fixes 😄 🎉

**Pull request merged**